### PR TITLE
Warn on cdist p-norm overflow for small p

### LINKED ISSUE #187179: torch.cdist returns +Inf for small p (p-norm overflow, docstring admits p in [0, inf])
### 🐛 Describe the bug

`torch.cdist(x1, x2, p)` evaluates the p-norm directly as `((|x1 - x2|)^p).sum(-1)^(1/p)`. For small `p` on unit-scale input each `(|diff|)^p ~= 1`, so the per-pair sum is `~= M` (the feature dim) and the outer `M^(1/p)` overflows fp32 to `+inf` once `(1/p) * log(M)` exceeds `log(3.4e38) ~= 88.7`. There is no validation on `p`, and the docstring explicitly admits the full range `p in [0, inf]`.

The threshold is much lower than absurd values. For `M = 4` overflow begins at `p ~= 0.027`; for `M = 768` (BERT-scale embeddings) at `p ~= 0.043`. A fractional-norm sweep like `p in {0.01, 0.1, 1, 2}` silently returns all `+inf` for the small-`p` end.

```python
import torch

torch.manual_seed(0)
x1 = torch.rand([2, 3, 4])
x2 = torch.rand([2, 5, 4])
for p in [0.0001, 0.001, 0.01, 0.1, 1.0]:
    out = torch.cdist(x1, x2, p=p)
    print(f"p={p}: finite_frac={torch.isfinite(out).float().mean().item()}")
```

Observed:

```
p=0.0001: finite_frac=0.0
p=0.001: finite_frac=0.0
p=0.01: finite_frac=0.0
p=0.1: finite_frac=1.0
p=1.0: finite_frac=1.0
```

Expected:

Either (a) raise `ValueError` (or warn) when `(1/p) * log(M)` would exceed `log(dtype_max)`, or (b) evaluate the p-norm in log space via `exp((1/p) * logsumexp(p * log(|diff| + eps)))`, finite for any `p > 0`. Same overflow family as `torch.nn.PairwiseDistance` (#184036, large `p`) and `torch.nn.functional.triplet_margin_loss` (small `p`).

### Versions

```
PyTorch version: 2.5.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.10.20 (main, Mar 11 2026, 17:46:40) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-124-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1660 SUPER
Nvidia driver version: 535.309.01
cuDNN version: Could not collect
Is XNNPACK available: True

CPU:
Architecture:                            x86_64
CPU(s):                                  20
Model name:                              12th Gen Intel(R) Core(TM) i7-12700
Thread(s) per core:                      2
Core(s) per socket:                      12

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] torch==2.5.1+cu121
[pip3] triton==3.1.0
```

### PR DESCRIPTION
**Pull request by wasimat404**
Fixes #187179

### Summary

`torch.cdist(x1, x2, p)` with `0 < p < 1` silently returns `+inf` on
wide inputs. The p-norm scales as `M^(1/p)` in the feature dimension `M`,
which diverges as `p -> 0` and overflows the dtype once `(1/p)*log(M)`
exceeds `log(max_float)`. For `M=768` (BERT-scale) this hits at
`p ~= 0.043`, well inside the range the docstring admits (`p in [0, inf]`).

### Why a warning rather than a numerical rewrite

The issue suggests a log-space form `exp((1/p)*logsumexp(p*log|diff|))`.
That fixes the *large*-p sibling (#184036) but not this case: for small
`p` it collapses to `M^(1/p)` and overflows at the final `exp`. The same
is true of max-factoring. The result genuinely exceeds the representable
range, so no rescaling keeps it finite — a warning is the correct fix.

I root-caused the overflow to the `pdist_calc` functor in
`cdist_kernel_impl` (CPU + CUDA); the divergence is mathematical, not a
kernel precision issue.

### Implementation

A guard in `cdist_impl` warns when `(1/p)*log(M)` would exceed
`log(max_float)`. It only fires when `M` is statically known and
`p in (0, 1)`; all other paths are untouched and pay no cost.

### Tests

- `test_cdist_small_p_overflow_warns` — warning fires (small p, wide M)
- `test_cdist_normal_p_no_overflow_warn` — silent for normal p

Both pass; `lintrunner` clean.

### CODE REVIEW

**Comment by wasimat404** [comment]
cc @lezcano @nikitaved  this is the small-p half of the cdist/PairwiseDistance overflow family (#187179, related to #184036).

**Comment by albanD** [comment]
Thank you very much for sending a PR.
To avoid you spending time preparing a PR for a feature we're not sure we want to add, our usual process is to discuss on the issue before doing the PR.
In particular, make sure that the issue is marked "actionable" before sending a PR!
See https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions for details!
